Editorial review: Document CSS carousel functionality (#38752)
mdn/content: files/en-us/web/css/_doublecolon_scroll-button/index.md

---
title: ::scroll-button()
slug: Web/CSS/::scroll-button
page-type: css-pseudo-element
browser-compat: css.selectors.scroll-button
---

{{CSSRef}}

The **`::scroll-button()`** [CSS](/en-US/docs/Web/CSS) [pseudo-element](/en-US/docs/Web/CSS/Pseudo-elements) represents a button for controlling the scrolling of a {{glossary("scroll container")}}.

## Syntax

```css-nolint
::scroll-button(<scroll-button-direction>) {
  /* ... */
}
```

### Parameters

- `<scroll-button-direction>`

  - : A value representing which direction of scroll button you want to select. The following values are available:

    - `*`
      - : Selects all the originating element's scroll buttons, allowing styles to be applied to each of them in a single rule.
    - `down`
      - : Selects the button that will scroll the content downwards.
    - `left`
      - : Selects the button that will scroll the content left.
    - `right`
      - : Selects the button that will scroll the content right.
    - `up`

>       - : Selects the button that will scroll the content upwards.

Author: Updated to "upward"

The `::scroll-button()` pseudo-elements are generated inside a {{glossary("scroll container")}}, only when their {{cssxref("content")}} properties are set to a value other than `none`. They are generated as siblings of the scroll container's child DOM elements, immediately preceding them and any {{cssxref("::scroll-marker-group")}} generated on the container.

You can generate up to four scroll buttons per scroll container, which will scroll the content towards the start and end of the block and inline axes. The selector's argument specifies which scrolling direction is selected. You can also specify a value of `*` to target all of the `::scroll-button()` pseudo-elements, providing styles to all the buttons in a single rule.

The generated buttons behave just like regular {{htmlelement("button")}} elements, including sharing their default browser styles. They are focusable, accessible, and can be activated like regular buttons. When a scroll button is pressed, the scroll container's content is scrolled in the specified direction by one "page," or approximately the dimension of the scroll container, similar to pressing <kbd>PgUp</kbd> and <kbd>PgDn</kbd> keys.

> [!NOTE]

> > Because the buttons do not exist in the DOM like regular buttons, they cannot be directly targeted with JavaScript.

Author: Agreed; deleted.

}
```

> [!NOTE]
> We also set a {{cssxref("cursor")}} value of `pointer` on the scroll buttons to make it more obvious that they can be interacted with (an improvement for both general [UX](/en-US/docs/Glossary/UX) and [cognitive accessibility](/en-US/docs/Web/Accessibility/Guides/Cognitive_accessibility)), unsetting it when the scroll buttons are `:disabled`.

Next, an appropriate icon is set on the left and right scroll buttons via the `content` property, which is also what causes the scroll buttons to be generated:

```css live-sample___creating-scroll-buttons live-sample___positioning-scroll-buttons
ul::scroll-button(left) {
  content: "◄";
}

ul::scroll-button(right) {
  content: "►";
}
```

#### Result

{{EmbedLiveSample("creating-scroll-buttons", '', '220')}}

Note how the scroll buttons are created at the bottom left on the carousel. Try pressing them to see how they cause the content to be scrolled.

### Positioning the scroll buttons

The above example works, but the buttons are not ideally placed. In this section we will add some more CSS to position them appropriately using [CSS anchor positioning](/en-US/docs/Web/CSS/CSS_anchor_positioning)..

#### CSS

> First of all, a reference {{cssxref("anchor-name")}} is set on the list. Next, each scroll button has its {{cssxref("position")}} set to `absolute`, and its {{cssxref("position-anchor")}} property set to the same reference name defined on the list, to associate the two together.

Author: Overly wordy, plus this implies that the buttons and list are anchor-positioned at this point, which they are not (although the linked section would clear this up when followed). I've gone for something a bit simpler:

> First of all, a reference {{cssxref("anchor-name")}} is set on the `<ul>` to define it as a named anchor. Next, each scroll button has its {{cssxref("position")}} set to `absolute` and its {{cssxref("position-anchor")}} property set to the list's `anchor-name`, to [associate the two together](/en-US/docs/Web/CSS/CSS_anchor_positioning/Using#associating_anchor_and_positioned_elements).

## Syntax

```css-nolint
::scroll-button(<scroll-button-direction>) {
  /* ... */
}
```

### Parameters

- `<scroll-button-direction>`

  - : A value representing which direction of scroll button you want to select. The following values are available:

    - `*`
      - : Selects all four scroll buttons, allowing styles to be applied to each of them in a single rule.
    - `down`
      - : Selects the downwards scroll button.
    - `left`
      - : Selects the leftwards scroll button.
    - `right`
      - : Selects the rightwards scroll button.
    - `up`
      - : Selects the upwards scroll button.
    - `block-end`
      - : Selects the button that will scroll the content in the block-end direction.
    - `block-start`
      - : Selects the button that will scroll the content in the block-start direction.
    - `inline-end`
      - : Selects the button that will scroll the content in the inline-end direction.
    - `inline-start`
      - : Selects the button that will scroll the content in the inline-start direction.

    The specification also defines two other values — `next` and `prev` — but these are not currently supported in any browser.

## Description

> The `::scroll-button()` pseudo-elements are generated inside a {{glossary("scroll container")}}, only when their {{cssxref("content")}} properties are set to a value other than `none`. They are generated as siblings of the scroll container's child DOM elements, immediately preceding them, but after a generated {{cssxref("::scroll-marker-group")}} pseudo-element.

Author: OK, I agree. Deleted the last bit of the sentence.

> Which button is selected is determined by the selector's argument, which specifies which scrolling direction you want to select the button for. You can generate up to four scroll buttons per scroll container, which will scroll the content towards the start and end of the block and inline axes. You can also specify a value of `*` to target all of the `::scroll-button()` pseudo-elements with styles in a single rule.

Author: Nice. Updated.

> The generated buttons behave just like regular {{htmlelement("button")}} elements, including sharing their default browser styles. They are focusable, etc. When a scroll button is pressed, the scroll container's content is scrolled in the specified direction by an amount determined by the browser. The recommendation is to set up [CSS scroll snapping](/en-US/docs/Web/CSS/CSS_scroll_snap) on the scroll container and set each separate item of content you want to scroll to as a [snap target](/en-US/docs/Glossary/Scroll_snap#snap_target). This being the case, the buttons will scroll the content to a new snap target each time they are pressed. While the scroll buttons will work without scroll snapping, you might not get the desired effect.

Reviewer: ```suggestion
The generated buttons behave just like regular {{htmlelement("button")}} elements, including sharing their default browser styles. They are focusable, accessible, and can be activated like regular buttons. When a scroll button is pressed, the scroll container's content is scrolled in the specified direction by one "page," or approximately the dimension of the scroll container, in the direction set by the parameter, similar to pressing <kbd>PgUp</kbd> and <kbd>PgDn</kbd> keys. 

The recommendation is to set up [CSS scroll snapping](/en-US/docs/Web/CSS/CSS_scroll_snap) on the scroll container and set each separate item of content you want to scroll to as a [snap target](/en-US/docs/Glossary/Scroll_snap#snap_target). This being the case, activating a scroll button will scroll the content to the snap target that is one "page" away.  While the scroll buttons will work without scroll snapping, you might not get the desired effect.
```

Author: updated

> When it is not possible to scroll any further in a particular scroll button's scrolling direction, it is automatically disabled, otherwise it is enabled. You can style the scroll buttons in their enabled and disabled states using the {{cssxref(":enabled")}} and {{cssxref(":disabled")}} pseudo-classes.

Reviewer: ```suggestion
When it is not possible to scroll any further in a particular scroll button's scrolling direction, the button is automatically disabled, otherwise it is enabled. You can style the scroll buttons in their enabled and disabled states using the {{cssxref(":enabled")}} and {{cssxref(":disabled")}} pseudo-classes.
```

---
title: ::scroll-button()
slug: Web/CSS/::scroll-button
page-type: css-pseudo-element
browser-compat: css.selectors.scroll-button
---

{{CSSRef}}

The **`::scroll-button()`** [CSS](/en-US/docs/Web/CSS) [pseudo-element](/en-US/docs/Web/CSS/Pseudo-elements) represents a button for controlling the scrolling of a {{glossary("scroll container")}}.

## Syntax

```css-nolint
::scroll-button(<scroll-button-direction>) {
  /* ... */
}
```

### Parameters

- `<scroll-button-direction>`

  - : A value representing which direction of scroll button you want to select. The following values are available:

    - `*`

>       - : Selects all four scroll buttons, allowing styles to be applied to each of them in a single rule.

Author: updated

    - `down`

>       - : Selects the downwards scroll button.

Author: done

    - `left`
      - : Selects the leftwards scroll button.
    - `right`
      - : Selects the rightwards scroll button.
    - `up`
      - : Selects the upwards scroll button.
    - `block-end`
      - : Selects the button that will scroll the content in the block-end direction.
    - `block-start`
      - : Selects the button that will scroll the content in the block-start direction.
    - `inline-end`
      - : Selects the button that will scroll the content in the inline-end direction.
    - `inline-start`
      - : Selects the button that will scroll the content in the inline-start direction.

    The specification also defines two other values — `next` and `prev` — but these are not currently supported in any browser.

## Description

> The `::scroll-button()` pseudo-elements are generated inside a {{glossary("scroll container")}}, only when their {{cssxref("content")}} properties are set to a value other than `none`. They are generated as siblings of the scroll container's child DOM elements, immediately preceding them.

Author: I've updated the paragraph to correct it.

I had a bit of a look at JS selection of these buttons, and you are right, they can't be selected by `querySelector()`, or referenced using something like list.childNodes[x]. They don't really exist in the DOM, as such, afaics.

I've added a note in the description:

> [!NOTE]
> Because the buttons do not exist in the DOM like regular buttons, they cannot be directly targeted with JavaScript.

> The selector's argument specifies which scrolling direction is selected. You can generate up to four scroll buttons per scroll container, which will scroll the content towards the start and end of the block and inline axes. You can also specify a value of `*` to target all of the `::scroll-button()` pseudo-elements, providing styles to all the buttons in a single rule.

Author: updated

The recommendation is to set up [CSS scroll snapping](/en-US/docs/Web/CSS/CSS_scroll_snap) on the scroll container and set each separate item of content you want to scroll to as a [snap target](/en-US/docs/Glossary/Scroll_snap#snap_target). This being the case, activating a scroll button will scroll the content to the snap target that is one "page" away. While the scroll buttons will work without scroll snapping, you might not get the desired effect.

When it is not possible to scroll any further in a particular scroll button's scrolling direction, the button is automatically disabled, otherwise it is enabled. You can style the scroll buttons in their enabled and disabled states using the {{cssxref(":enabled")}} and {{cssxref(":disabled")}} pseudo-classes.

## Examples

### Creating carousel scroll buttons

> This demo creates a responsive, CSS-only carousel. It uses the {{cssxref("columns")}} property and the `::columns` pseudo-element to create arbitary content columns that span the full width of their parent {{glossary("scroll container")}}, which can be scrolled between using scroll buttons. Each column contains multiple items of content, which can vary in number depending on the viewport width.

Author: I've updated the example to follow the simpler thing we already did on the `::scroll-marker` page. I've included two stages — creating the buttons and positioning the buttons. I don't think a separate stage for styling the buttons is necessary.

BTW, I didn't experience a z-index problem.

}
```

> We don't need to set [alternative text](/en-US/docs/Web/CSS/content#alternative_text_string_counter) for the icons on the `content` as the browser takes care of providing appropriate accessible names automatically.

Reviewer: ```suggestion
We don't need to set [alternative text](/en-US/docs/Web/CSS/content#alternative_text_string_counter) for the icons on the `content` as the browser takes care of providing appropriate {{glossary("accessible name", "accessible names")}} automatically.
```

Author: done!
